### PR TITLE
Query: Use correct column mappings for TVF/FromSql

### LINKED ISSUE #21511: Query: ColumnMappings to use for TVF/FromSql
SelectExpression has a ctor (also associated public method in SqlExpressionFactory) which takes the TableExpressionBase which acts as the source of the data. This is used for scenarios like TVF/FromSql (which currently works with TPH or non-hierarchy only).
We need to determine which column mappings to use when generating columns from database, since this code path is encountered while not using any particular entity type mapping from model.

### PR DESCRIPTION
**Pull request by smitpatel**
Resolves #21511

### CODE REVIEW

**Comment by smitpatel** [comment]
@msftbot require 1 approval

File: src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs [hunk 1]
```diff
@@ -240,23 +239,42 @@ internal SelectExpression(IEntityType entityType, ISqlExpressionFactory sqlExpre
...
         }
 
         internal SelectExpression(IEntityType entityType, TableExpressionBase tableExpression)
             : base(null)
         {
+            ITableBase table;
+
+            switch (tableExpression)
+            {
+                case TableExpression tableExpression1:
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs:251]: Rename the other one to tableExpressionBase

File: src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs [hunk 2]
```diff
@@ -240,23 +239,42 @@ internal SelectExpression(IEntityType entityType, ISqlExpressionFactory sqlExpre
...
+                    table = entityType.GetViewOrTableMappings().FirstOrDefault(
+                        m => m.Table.Schema == tableExpression1.Schema
+                            && m.Table.Name == tableExpression1.Name).Table;
+                    break;
+
+                case FromSqlExpression fromSqlExpression:
+                    table = entityType.GetSqlQueryMappings().FirstOrDefault(m => m.IsDefaultSqlQueryMapping).Table;
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs:258]: Unnamed FromSql should use `GetDefaultMappings`. `SqlQueryMapping` with `IsDefaultSqlQueryMapping` set to `true` is only when the root is an entity type mapped using `ToQuerySql`

File: src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs [hunk 3]
```diff
@@ -240,23 +239,38 @@ internal SelectExpression(IEntityType entityType, ISqlExpressionFactory sqlExpre
...
+                    break;
+
+                case TableValuedFunctionExpression tableValuedFunctionExpression:
+                    table = entityType.GetFunctionMappings().SingleOrDefault(
+                        fm => fm.Table.Schema == tableValuedFunctionExpression.Schema
+                            && fm.Table.Name == tableValuedFunctionExpression.Name).Table;
+                    break;
+
+                default:
+                    table = entityType.GetDefaultMappings().Single().Table;
```
Comment (AndriySvyryd) [inline src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs:264]: Would be cleaner if `TableExpressionBase` had an `ITableBase` property

Comment (smitpatel) [reply src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs:264]: Select/Joins etc also are TableExpressionBase. They don't correspond to any ITableBase.
TableExpressionBase is anything from which you can project out columns. ITableBase correspond to actual database object only so we would need some middle type to hold it.

Comment (AndriySvyryd) [reply src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs:264]: The `ITableBase` property could be `null` if there's no database object

Comment (smitpatel) [reply src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs]: I added TableExpressiob.Table & TVFExpression.StoreFunction.
I will revisit if FromSqlExpression should take ITableBase when I work in ToQuerySql API.

**Pull request merged**